File: dev-settings/coding-agents/sync-cmd/internal/dirsync/dirsync.go
package dirsync

import (
	"io"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
)

func Sync(src, dst string) error {
	if info, err := os.Stat(dst); err == nil && info.IsDir() {
		if err := deleteExtraFiles(src, dst); err != nil {
			return err
		}
		if err := deleteEmptyDirs(dst); err != nil {
			return err
		}
	}

	return copyTree(src, dst)
}

func CopyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	info, err := in.Stat()
	if err != nil {
		return err
	}

	out, err := os.OpenFile(dst, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, info.Mode())
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

func deleteExtraFiles(src, dst string) error {
	return filepath.WalkDir(dst, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() {
			return nil
		}
		rel, _ := filepath.Rel(dst, path)
		srcPath := filepath.Join(src, rel)
		if _, err := os.Stat(srcPath); os.IsNotExist(err) {
			return os.Remove(path)
		}
		return nil
	})
}

func deleteEmptyDirs(dst string) error {
	var dirs []string
	filepath.WalkDir(dst, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() && path != dst {
			dirs = append(dirs, path)
		}
		return nil
	})

	sort.Slice(dirs, func(i, j int) bool {
		return len(filepath.SplitList(dirs[i])) > len(filepath.SplitList(dirs[j]))
	})

	for _, d := range dirs {
		entries, err := os.ReadDir(d)
		if err != nil {
			continue
		}
		if len(entries) == 0 {
			os.Remove(d)
		}
	}
	return nil
}

func copyTree(src, dst string) error {
	return filepath.WalkDir(src, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		rel, _ := filepath.Rel(src, path)
		target := filepath.Join(dst, rel)

		if d.IsDir() {
			return os.MkdirAll(target, 0755)
		}

		return CopyFile(path, target)
	})
}

File: dev-settings/coding-agents/sync-cmd/sync_cmd/dirsync.py
import shutil
from pathlib import Path


def sync(src: Path, dst: Path) -> None:
    src, dst = Path(src), Path(dst)
    if dst.is_dir():
        _delete_extra_files(src, dst)
        _delete_empty_dirs(dst)
    _copy_tree(src, dst)


def copy_file(src: Path, dst: Path) -> None:
    src, dst = Path(src), Path(dst)
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(src, dst)


def _delete_extra_files(src: Path, dst: Path) -> None:
    for path in list(dst.rglob("*")):
        if path.is_file():
            rel = path.relative_to(dst)
            if not (src / rel).exists():
                path.unlink()


def _delete_empty_dirs(dst: Path) -> None:
    dirs = sorted(
        [p for p in dst.rglob("*") if p.is_dir() and p != dst],
        key=lambda d: len(d.parts),
        reverse=True,
    )
    for d in dirs:
        if not any(d.iterdir()):
            d.rmdir()


def _copy_tree(src: Path, dst: Path) -> None:
    for path in src.rglob("*"):
        rel = path.relative_to(src)
        target = dst / rel
        if path.is_dir():
            target.mkdir(parents=True, exist_ok=True)
        else:
            target.parent.mkdir(parents=True, exist_ok=True)
            shutil.copy2(path, target)
